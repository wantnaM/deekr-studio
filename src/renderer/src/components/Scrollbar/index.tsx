import { throttle } from 'lodash'
import { FC, useCallback, useEffect, useRef, useState } from 'react'
import styled from 'styled-components'

interface Props extends Omit<React.HTMLAttributes<HTMLDivElement>, 'onScroll'> {
  ref?: React.RefObject<HTMLDivElement | null>
  onScroll?: () => void // Custom onScroll prop for useScrollPosition's handleScroll
}

const Scrollbar: FC<Props> = ({ ref: passedRef, children, onScroll: externalOnScroll, ...htmlProps }) => {
  const [isScrolling, setIsScrolling] = useState(false)
  const timeoutRef = useRef<NodeJS.Timeout | null>(null)

  const clearScrollingTimeout = useCallback(() => {
    if (timeoutRef.current) {
      clearTimeout(timeoutRef.current)
      timeoutRef.current = null
    }
  }, [])

  const handleScroll = useCallback(() => {
    setIsScrolling(true)
    clearScrollingTimeout()
    timeoutRef.current = setTimeout(() => {
      setIsScrolling(false)
      timeoutRef.current = null
    }, 1500)
  }, [clearScrollingTimeout])

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const throttledInternalScrollHandler = useCallback(throttle(handleScroll, 100, { leading: true, trailing: true }), [
    handleScroll
  ])

  // Combined scroll handler
  const combinedOnScroll = useCallback(() => {
    throttledInternalScrollHandler()
    if (externalOnScroll) {
      externalOnScroll()
    }
  }, [throttledInternalScrollHandler, externalOnScroll])

  useEffect(() => {
    return () => {
      clearScrollingTimeout()
      throttledInternalScrollHandler.cancel()
    }
  }, [throttledInternalScrollHandler, clearScrollingTimeout])

  return (
    <Container
      {...htmlProps} // Pass other HTML attributes
      $isScrolling={isScrolling}
      onScroll={combinedOnScroll} // Use the combined handler
      ref={passedRef}>
      {children}
    </Container>
  )
}

const Container = styled.div<{ $isScrolling: boolean }>`
  overflow-y: auto;
  &::-webkit-scrollbar-thumb {
    transition: background 2s ease;
    background: ${(props) => (props.$isScrolling ? 'var(--color-scrollbar-thumb)' : 'transparent')};
    &:hover {
      background: var(--color-scrollbar-thumb-hover);
    }
  }
`

Scrollbar.displayName = 'Scrollbar'

export default Scrollbar
