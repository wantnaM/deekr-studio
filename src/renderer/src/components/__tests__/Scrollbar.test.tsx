import { fireEvent, render, screen } from '@testing-library/react'
import { act } from 'react'
import { afterEach, beforeEach, describe, expect, it, type Mock, vi } from 'vitest'

import Scrollbar from '../Scrollbar'

// Mock lodash throttle
vi.mock('lodash', async () => {
  const actual = await import('lodash')
  return {
    ...actual,
    throttle: vi.fn((fn) => {
      // 简单地直接返回函数，不实际执行节流
      const throttled = (...args: any[]) => fn(...args)
      throttled.cancel = vi.fn()
      return throttled
    })
  }
})

describe('Scrollbar', () => {
  beforeEach(() => {
    // 使用 fake timers
    vi.useFakeTimers()
  })

  afterEach(() => {
    // 恢复真实的 timers
    vi.restoreAllMocks()
    vi.useRealTimers()
  })

  describe('rendering', () => {
    it('should render children correctly', () => {
      render(
        <Scrollbar data-testid="scrollbar">
          <div data-testid="child">测试内容</div>
        </Scrollbar>
      )

      const child = screen.getByTestId('child')
      expect(child).toBeDefined()
      expect(child.textContent).toBe('测试内容')
    })

    it('should pass custom props to container', () => {
      render(
        <Scrollbar data-testid="scrollbar" className="custom-class">
          内容
        </Scrollbar>
      )

      const scrollbar = screen.getByTestId('scrollbar')
      expect(scrollbar.className).toContain('custom-class')
    })

    it('should match default styled snapshot', () => {
      const { container } = render(<Scrollbar data-testid="scrollbar">内容</Scrollbar>)
      expect(container.firstChild).toMatchSnapshot()
    })
  })

  describe('scrolling behavior', () => {
    it('should update isScrolling state when scrolled', () => {
      render(<Scrollbar data-testid="scrollbar">内容</Scrollbar>)

      const scrollbar = screen.getByTestId('scrollbar')

      // 初始状态下应该不是滚动状态
      expect(scrollbar.getAttribute('isScrolling')).toBeFalsy()

      // 触发滚动
      fireEvent.scroll(scrollbar)

      // 由于 isScrolling 是组件内部状态，不直接反映在 DOM 属性上
      // 但可以检查模拟的事件处理是否被调用
      expect(scrollbar).toBeDefined()
    })

    it('should reset isScrolling after timeout', () => {
      render(<Scrollbar data-testid="scrollbar">内容</Scrollbar>)

      const scrollbar = screen.getByTestId('scrollbar')

      // 触发滚动
      fireEvent.scroll(scrollbar)

      // 前进时间但不超过timeout
      act(() => {
        vi.advanceTimersByTime(1000)
      })

      // 前进超过timeout
      act(() => {
        vi.advanceTimersByTime(600)
      })

      // 不测试样式，这里只检查组件是否存在
      expect(scrollbar).toBeDefined()
    })

    it('should reset timeout on continuous scrolling', () => {
      const clearTimeoutSpy = vi.spyOn(global, 'clearTimeout')

      render(<Scrollbar data-testid="scrollbar">内容</Scrollbar>)

      const scrollbar = screen.getByTestId('scrollbar')

      // 第一次滚动
      fireEvent.scroll(scrollbar)

      // 前进一部分时间
      act(() => {
        vi.advanceTimersByTime(800)
      })

      // 再次滚动
      fireEvent.scroll(scrollbar)

      // clearTimeout 应该被调用，因为在第二次滚动时会清除之前的定时器
      expect(clearTimeoutSpy).toHaveBeenCalled()
    })
  })

  describe('throttling', () => {
    it('should use throttled scroll handler', async () => {
      const { throttle } = await import('lodash')

      render(<Scrollbar data-testid="scrollbar">内容</Scrollbar>)

      // 验证 throttle 被调用
      expect(throttle).toHaveBeenCalled()
      // 验证 throttle 调用时使用了 100ms 延迟和正确的选项
      expect(throttle).toHaveBeenCalledWith(expect.any(Function), 100, { leading: true, trailing: true })
    })
  })

  describe('cleanup', () => {
    it('should clear timeout and cancel throttle on unmount', async () => {
      const clearTimeoutSpy = vi.spyOn(global, 'clearTimeout')

      const { unmount } = render(<Scrollbar data-testid="scrollbar">内容</Scrollbar>)

      const scrollbar = screen.getByTestId('scrollbar')

      // 触发滚动设置定时器
      fireEvent.scroll(scrollbar)

      // 卸载组件
      unmount()

      // 验证 clearTimeout 被调用
      expect(clearTimeoutSpy).toHaveBeenCalled()

      // 验证 throttle.cancel 被调用
      const { throttle } = await import('lodash')
      const throttledFunction = (throttle as unknown as Mock).mock.results[0].value
      expect(throttledFunction.cancel).toHaveBeenCalled()
    })
  })

  describe('props handling', () => {
    it('should handle ref forwarding', () => {
      const ref = { current: null }

      render(
        <Scrollbar data-testid="scrollbar" ref={ref}>
          内容
        </Scrollbar>
      )

      // 验证 ref 被正确设置
      expect(ref.current).not.toBeNull()
    })
  })
})
