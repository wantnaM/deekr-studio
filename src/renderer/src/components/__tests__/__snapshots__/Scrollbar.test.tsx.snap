// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Scrollbar > rendering > should match default styled snapshot 1`] = `
.c0 {
  overflow-y: auto;
}

.c0::-webkit-scrollbar-thumb {
  transition: background 2s ease;
  background: transparent;
}

.c0::-webkit-scrollbar-thumb:hover {
  background: var(--color-scrollbar-thumb-hover);
}

<div
  class="c0"
  data-testid="scrollbar"
>
  内容
</div>
`;
